[POS] Custom Payment UI — Part 3 | Emitting UI–agnostic payment states in `CardReaderPaymentController`

samiuelson opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #12825
<!-- Id number of the GitHub issue this PR addresses. -->

#### This is **part 3** of 5 PRs refactoring the Payment flow:
1. [[POS] Custom payment UI — Part 1 | Separating payment collection management from Android Framework](https://github.com/woocommerce/woocommerce-android/pull/12853)
2. [[POS] Custom payment UI — Part 2 | Separating CardReaderPaymentController's events from MultiLiveEvent](https://github.com/woocommerce/woocommerce-android/pull/12870)
3. **[[POS] Custom Payment UI – Part 3 | Emitting UI–agnostic payment states in CardReaderPaymentController](https://github.com/woocommerce/woocommerce-android/pull/12877)**
4. [[POS] Custom payment UI — Part 4 | Switch from viewStateData to paymentState](https://github.com/woocommerce/woocommerce-android/pull/12898)
5. [[POS] Custom payment UI — Part 5 | Unit tests clean up](https://github.com/woocommerce/woocommerce-android/pull/12943)

⚠️ Don't merge — the branch will be merged together with the above ones after additional testing of the whole refactor.

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR introduces a new class modeling a "payment state" — `CardReaderPaymentOrRefundState`. It is emitted by the `paymentState: StateFlow<CardReaderPaymentOrRefundState>` property in `CardReaderPaymentController`, allowing observing the payment state in POS.

The goal of the `paymentState` property is to replace the `viewStateData: LiveData<ViewState>` that emits "UI model" instances, designed specifically for the existing dialog-based payment flow UI. For now, the `viewStateData` is marked deprecated, and supposed to be used only by the existing IPP flow.

#### Coming up in the next PR:
1. Removing viewStateData from the controller
2. Mapping the controller's `paymentState` to `ViewState` in `CardReaderPaymentViewModel`
3. Covering controller with unit tests

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->
As a result of the refactoring done within this PR, the app should work without any change. It's crucial to test the IPP flow in the store management and POS modes against regression. It may be useful to base on the test plan (pdfdoF-5Jz-p2).

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->
This PR will be followed up by removing the `viewStateData` property from the controller, adding unit tests, and mapping the `paymentState` to `ViewState`. That said, this PR should not cause any regression in the existing IPP flow, however, the most important part to test is the code changes.

- [ ] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Introduce `CardReaderPaymentState` model emitted by `CardReaderPaymentController`

The goal is to remove viewStateData property and map paymentState to ViewState in the target [ViewModel]

samiuelson committed:
Introduce `CardReaderPaymentState` model

The goal is to make payment controller return UI-agnostic payment state and map the payment state to UI state in the target VM.

samiuelson committed:
Map payment flow error to `PaymentFailed` state

samiuelson committed:
Emit `FetchingOrderFailed` state as `paymentState`

dangermattic commented:
<!--
  3 Errors: Please add tests for class <co...
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          3 Errors
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>CardReaderPaymentOrRefundState</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>CardReaderPaymentState</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">This PR is tagged with <code>status: do not merge</code> label(s).</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

samiuelson committed:
Emit `LoadingData` state as `paymentState`

samiuelson committed:
Emit `CollectingPayment` state as `paymentState`

samiuelson committed:
Emit `ProcessingPayment` state as `paymentState`

samiuelson committed:
Emit `CapturingPayment` state as `paymentState`

samiuelson committed:
Fix detekt's issues

samiuelson committed:
Emit `ReFetchingOrder` as `paymentState`

samiuelson committed:
Emit `SharingReceipt` as `paymentState`

samiuelson committed:
Emit `PrintingReceipt` as `paymentState`

samiuelson committed:
Emit `PaymentSuccessful` as `paymentState`

samiuelson committed:
Emit remaining states as `paymentState`

samiuelson committed:
Format code

samiuelson committed:
Clean up code

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/payment/controller/CardReaderPaymentController.kt:
```diff
@@ -127,8 +130,20 @@ class CardReaderPaymentController(
+     *
+     * Returns observable of [CardReaderPaymentOrRefundState].
+     */
+    val paymentState: StateFlow<CardReaderPaymentOrRefundState> = _paymentState
```
❓For now, I ended up leaving the existing `viewStateData: LiveData<ViewState>` property and emitting the new one in parallel—the `paymentState: StateFlow<CardReaderPaymentOrRefundState>`. 

The `paymentState` emits a newly introduced "payment state" model which is independent of the UI (unlike the old `viewStateData`), and will be used by `WooPosTotalsViewModel` to calculate UI state. The `viewStateData` is marked deprecated and is supposed to be used only by the existing IPP flow.

Ideally, we would remove the `viewStateData` from `CardReaderPaymentController` and map `paymentState` to `viewStateData: LiveData<ViewState>` inside the `CardReaderPaymentViewModel`. 

I'm not sure though if going that next step further is justified, and would like to avoid too much interference/modification in the existing payment flow codebase. 

Let me know what you think @kidinov, @malinajirka, and @AnirudhBhat!

malinajirka replied:
Thanks for the ping @samiuelson! 

> Ideally, we would remove the viewStateData from CardReaderPaymentController and map paymentState to viewStateData: LiveData<ViewState> inside the CardReaderPaymentViewModel.

I'd personally consider creating a separate mapper class which would map `CardReaderPaymentOrRefundState` to `ViewState` (Or the other way round). It seems that it'd be straightforard and we'd avoid having to duplicate the calls here. Or am I missing something and it wouldn't be simple? Putting it inside the ViewModel would be an option as well, but in my experience VMs tend to grow so anything that can be easily encapsulated would ideally live in its own class.

I have two concerns with the current approach:
1. Feels quite error prone, we can easily add one more update to `paymentState` forgetting to update also the `ViewState`.
2. The bigger concerns is the precedent it creates - someone in the future might put IPP specific logic into this class since one example is already there (ViewState).

Having said all that, I'm working with the assumption that creating the mapper is a matter of a couple hours + testing IPP which we need to test anyway after all these changes. So definitely take it with a grain of salt.

samiuelson replied:
Thanks for sharing your thoughts, @malinajirka!

> I'd personally consider creating a separate mapper class which would map CardReaderPaymentOrRefundState to ViewState. It seems that it'd be straightforard and we'd avoid having to duplicate the calls here. Or am I missing something and it wouldn't be simple.

Yes, it's feasible to map `CardReaderPaymentOrRefundState` to `ViewState` 👍🏼 

I agree that encapsulating the state mapping logic into a dedicated class is a good approach. I think the mappers can be used outside the `CardReaderPaymentController` though, to keep the controller simple. For example, defining the mapper something like this:
```kotlin
class Mapper {
  operator fun invoke(): (CardReaderPaymentOrRefundState)->ViewState = // ...
}
```
would allow mapping payment state in the VM like that:
```kotlin
class ViewModel {
  val viewStateData: LiveData<ViewState> = paymentController.paymentState.map(maper()).asLiveData(coroutineContext)
}
```
which I think won't clutter the VM too much. 

> I have two concerns with the current approach:

> 1. Feels quite error prone, we can easily add one more update to paymentState forgetting to update also the ViewState.
> 2. The bigger concerns is the precedent it creates - someone in the future might put IPP specific logic into this class since one example is already there (ViewState).

Good points 👍🏼 I agree the code is not maintainable. Thanks for pointing that out 🙇🏼 

### Summary

As the action point, in the additional/next PR I'm going to:
1. Remove `viewStateData` from controller
2. Map the controller's `paymentState` to `ViewState` in `CardReaderPaymentViewModel`
3. Cover controller with unit tests

malinajirka replied:
Sounds great, thank you!

samiuelson committed:
Fix detekt's complaints

samiuelson committed:
Update sut instances in unit test

samiuelson committed:
Revert interac tracking code

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/payment/controller/CardReaderPaymentOrRefundState.kt:
```diff
@@ -0,0 +1,182 @@
+import com.woocommerce.android.ui.payments.cardreader.payment.InteracRefundFlowError
+import com.woocommerce.android.ui.payments.cardreader.payment.PaymentFlowError
+
+sealed class CardReaderPaymentOrRefundState {
```
I like the idea of having a separate state here!

Some ideas I have for `maybe` an improvement

What I am not sure about is that it still contains things that do not really belong here:
* `nameForTracking` - maybe to extract `trackCancelledFlow` to a separate class and make a decision what to track based on the type there?
* Lyambdas - maybe it's better to call the controller methods from the VMs, instead of passing them inside of the state? It should be easier to unit test that, and in my opinion, the code will be more straightforward to follow. Overall, state probably should contain only data

malinajirka replied:
I second the tracking part - that doesn't feel related to the UI so IMHO shouldn't be part of the state.

Regarding lambdas, I agree with the code being more straightforward without them. On the other hand I like that with lambdas it's easy to check what actions are available on each state - especially here where each state is kind of a different screen. I'd also argue it's easier not harder to unit test => the main benefit is that you can be sure the UI layer is not invoking incorrect methods since it's invoking only the methods which are available in the state which are covered by unit tests. I don't mind which solution we opt for, just wanted to share my POV.

malinajirka replied:
One more argument for avoiding lambdas is that if we create two instances of stateA => they won't be equal because of the lambda (lambdas are equal only if we are comparing the same instance of the lambdas) => it might ultimately result in unnecessary recomposition I guess? (I have never dove into Compose to be sure about this, I'm sure about the not-equal part in kotlin but not sure about the the compose part - something to consider).

kidinov replied:
>  I'd also argue it's easier not harder to unit test => the main benefit is that you can be sure the UI layer is not invoking incorrect methods since it's invoking only the methods which are available in the state which are covered by unit tests.

To test lambda functions, we first need to configure the controller to emit a specific state using these lambdas. Then, we invoke the lambdas returned from that state. In contrast, if we're dealing with a public method, we simply call it as part of the test and then validate the emitted state afterward

I'm not sure how we can test whether only the correct methods are invoked. Just because a specific set of lambdas is returned doesn’t mean that other public methods or lambdas, which may return in a different state, haven't been called. We can use `verifyNoMoreInteractions`, but in many cases, there may be other valid interactions to consider. Overall, I'd say we should strive more to the black box testing: `method invoked` -> `validate that expected state emitted`, without validating that specific method was invoked/not invoked

> One more argument for avoiding lambdas is that if we create two instances of stateA => they won't be equal because of the lambda (lambdas are equal only if we are comparing the same instance of the lambdas) => it might ultimately result in unnecessary recomposition I guess?

My understanding is that those states that Samuel created here won't go to compose but will be mapped to the actual view states in the VM and they will contain specific strings, images etc. 

Our current approach is to have something like this in the VMs so user interactions are pushed as events from composes. So kinda 1 public method to handle all users actions

```
fun onUIEvent(event: WooPosTotalsUIEvent) {
```

If we still want to pass lyambdas to compose, if those are methods of the VM, then it should be fine. Although chat still gpt recommends `to be completely sure they don’t cause extra recompositions, it’s best to pass them directly without re-creating them in the composable function.`

malinajirka replied:
>To test lambda functions, we first need to configure the controller to emit a specific state using these lambdas.

I look at this as a benefit of lambdas - you can invoke it only in a state that can happen in production as oppose to using public methods which you can invoke any time in the test no matter the state of the VM.

>  if we're dealing with a public method, we simply call it as part of the test and then validate the emitted state afterward

This is true, but only if the VM doesn't hold any state, right? If it holds state, you either need to replicate the state before calling the method (applies even for lambdas) or you might be calling a method in a state that can't happen in production (can't happen with lambdas).

> I'm not sure how we can test whether only the correct methods are invoked.

This is the benefit of lambdas IMO, unless there is a lambda which invokes the method, you can't invoke it (or you technically can but it's obvious from the code you are invoking something you shouldn't be).

> Just because a specific set of lambdas is returned doesn’t mean that other public methods or lambdas, which may return in a different state, haven't been called.

I mean, you can always invoke something which you shouldn't be invoking as a developer - if we go into an extreme you can use reflection. But if we consider realistic scenarios, if a lambda is part of the state you can't easily invoke it unless it is the current state => that ensures you are invoking only the methods at the right moments without having to use `verifyNoInteractions` and similar (it also significantly reduces the universe of possible states and order of possible invocations).

> Overall, I'd say we should strive more to the black box testing: method invoked -> validate that expected state emitted, without validating that specific method was invoked/not invoked

I agree, I also prefer black box testing, but I find it easier to push for black box testing with lambdas than with all methods being publicly available and developers needing to know which one to invoke when. (I'm omitting scenarios in which we'd be able to have all functions as pure functions which should always be preferred if achievable, but in my experience it's not achievable in the MVVM - wdyt?). I also find it easier to write tests and find which tests are missing with lambdas being part of the state as you just test all the states (and lambdas which are part of that state) => if you cover that, you have covered all the possible states that might happen in production.



Naive example:
```
class VM {
boolean paymentInProgress

fun cancelPayment() {
   payment.cancel().also {paymentInProgress = false}
}

}
```

```
sealed class ViewState {
     Idle()
     PaymentInProgress(cancelLambda)
}
```

With lambdas, you reasonably can't call `cancelPayment` unless you are in `PaymentInProgress` state. Without lambdas, you can call it whenever and you need to handle scenarios when someone invokes it when it's not in the in-progress. In both cases, in unit tests you need to replicate the state of a payment being in progress before testing if `cancelPayment` works or doesn't work.

-----------------------------------

Just to be explicit, I'm not saying we should be using lambdas - I shared why [I personally like them here](https://github.com/woocommerce/woocommerce-android/pull/12877#discussion_r1836815577) and I'm just trying to provide more context and clarify why I believe some of the downsides you are mentioning I interpret as upsides or as not fully relevant assuming developers follow set principles. I still think both approaches are completely valid, each of them has some pros and cons. In an ideal scenario, we'd be able to create a set of pure functions which I'd 100% vote for, but I'm yet to see that implemented for a view-vm layer with a reasonable design/effort/maintenance.

samiuelson replied:
Thanks for brainstorming and analyzing the pros and cons @malinajirka and @kidinov. This is quite helpful! 

Wrapping up, I see the following approaches possible:

**1. Having lambdas in payment state only**

* UI states rendered in compose would not contain lambdas.
* UI events would be passed from Composable to VM using the existing mechanism: `onUIEvent(event: WooPosTotalsUIEvent)`. VM would invoke the lambda from the current payment state exposed by payment controller. 
* This would require casting the current payment state to the expected sub-type, however, since the UI state will be in sync with the payment state it should be feasible.

**2. Having lambdas both in payment state and UI states (passing lambdas to Compose)**

* This would require abandoning `onUIEvent(event: WooPosTotalsUIEvent)` mechanism for handling user actions.
* UI states would also require special tweaks and attention to make them stable. 
* This would require some time and effort (difficult to estimate).

**3. Avoid using lambdas—exposing public methods in controller for payment “cancel” and “retry” operations.**

* This would require implementing extra safety checks in the controller, e.g. throwing exceptions in case these methods should not be invoked at the current payment state. 
* No effective way to “hide” methods in certain states.


Having said that, unless you have objections, I opt for taking the first approach which seems optimal. An additional benefit is that it would be compatible with the existing dialog-based IPP UI model, and UI state provider, reducing modifications to the existing IPP flow.

Let me know what you think!

samiuelson replied:
> nameForTracking - maybe to extract trackCancelledFlow to a separate class and make a decision what to track based on the type there?

Good point, I'm going to extract `nameForTracking` from the model.

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/payment/controller/CardReaderPaymentStateProvider.kt:
```diff
@@ -0,0 +1,118 @@
+import com.woocommerce.android.ui.payments.cardreader.payment.controller.CardReaderPaymentOrRefundState.CardReaderPaymentState.ProcessingPayment
+import javax.inject.Inject
+
+class CardReaderPaymentStateProvider @Inject constructor() {
```
❤️

samiuelson committed:
Add `onCancel` prop to `CollectingInteracRefund`

samiuelson committed:
Merge branch 'custom-payment-ui-2' into custom-payment-ui-3

samiuelson committed:
Extract `nameForTracking` from the state model

samiuelson committed:
Fix tracking logic – add missing names

samiuelson committed:
Extract `::trackCancelledFlow` into use case class

samiuelson commented:
Thanks for the reviews @kidinov and @malinajirka 🙇🏼 

As discussed, we're going to keep the callbacks in the `CardReaderPaymentOrRefundState`. Also, the mapping of `CardReaderPaymentOrRefundState` to `ViewState`, and removing `viewStateData` from controller (bringing the controller to the final shape) is ready for review in the next PR: [[POS] Custom payment UI — Part 4 | Switch from viewStateData to paymentState](https://github.com/woocommerce/woocommerce-android/pull/12898).

In the current PR, I extracted `nameForTracking` from the payment state model (c26ed71, 7fe2991) and moved the logic of the `::trackCancelledFlow` function into a use case class (44115aa).

Let me know what you think!

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/payment/controller/CardReaderPaymentController.kt:
```diff
@@ -1080,16 +1080,22 @@ class CardReaderPaymentController(
+                if (nameForTracking == null) {
+                    WooLog.e(WooLog.T.CARD_READER, "Invalid state received")
+                } else {
+                    tracker.trackInteracRefundCancelled(nameForTracking)
```
Did you miss `trackPaymentCancelled`?

samiuelson replied:
Yes, fixed that in the next commit ([7fe2991](https://github.com/woocommerce/woocommerce-android/pull/12877/commits/7fe29916c6cb5e08295118a318c3f6d260f0227e))

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/payment/controller/CardReaderPaymentController.kt:
```diff
@@ -1080,16 +1080,22 @@ class CardReaderPaymentController(
+                    else -> null
+                }
+                if (nameForTracking == null) {
+                    WooLog.e(WooLog.T.CARD_READER, "Invalid state received")
```
Btw, I'd consider to track this case too, just instead of passing hard coded string we could pass state's name there

samiuelson replied:
I like the idea, but I would prefer to implement it out of this PR, and refactoring process because it would affect the existing business logic—it would require modifying unit tests.

Reported issue: https://github.com/orgs/woocommerce/projects/278/views/1?visibleFields=%5B%22Title%22%2C%22Assignees%22%2C%22Status%22%2C143176801%2C%22Linked+pull+requests%22%5D&pane=issue&itemId=87370857&issue=woocommerce%7Cwoocommerce-android%7C12930

samiuelson committed:
Merge branch 'custom-payment-ui-2' into custom-payment-ui-3

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/payment/controller/CardReaderPaymentController.kt:
```diff
@@ -857,16 +1052,15 @@ class CardReaderPaymentController(
-            viewState.value?.let { state ->
-                trackCancelledFlow(state)
-            }
+            trackCancelledFlow(_paymentState.value)
```
❓ It seems that previously we'd track the result only when `state` wasn't null, but now we don't handle the scenario when _paymentState is null. Is this something that was missed or was it an intentional change since it can't be null?

samiuelson replied:
Yes, the null check was `LiveData.value`-specific (or Java-specific to be more precise).

Even though the `viewStateData: LiveData<ViewState>` is declared for non-nullable type (`ViewState`), the `value` prop doesn't guarantee null safety and returns nullable type. 

The `paymentState.value` returns non-nullable type, matching type it is declared for (`paymentState: StateFlow<CardReaderPaymentOrRefundState>`) so no null check is needed.

malinajirka replied:
Ahh, I see - I forgot about this difference between state flow and live data, thanks!

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/payment/controller/CardReaderTrackCanceledFlow.kt:
```diff
@@ -0,0 +1,41 @@
+import com.woocommerce.android.util.WooLog
+import javax.inject.Inject
+
+class CardReaderTrackCanceledFlow @Inject constructor(
```
🔍 If I recall correctly, we usually call these types of classes with `invoke` method `Actions` instead of `Flows` (just to reduce risk of confusion with Kotlin Flows). Might be worth renaming, wdyt?

samiuelson replied:
Done in [0c4e9ec](https://github.com/woocommerce/woocommerce-android/pull/12877/commits/0c4e9ecd9955fdb198590ec6794774b89e46c450)

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/payment/controller/CardReaderTrackCanceledFlow.kt:
```diff
@@ -0,0 +1,41 @@
+                else -> null
+            }
+            if (nameForTracking == null) {
+                WooLog.e(WooLog.T.CARD_READER, "Invalid state received")
```
🔍  We might want to consider crashing in DEBUG mode to ensure we don't miss these types of bugs. Wdyt?

samiuelson replied:
I logged that idea into the _tracking enhancement_ [issue](https://github.com/woocommerce/woocommerce-android/issues/12930). 

Such a change would require the update of ~20 unit tests. Most of these tests verify if the proper terminal events are emitted, etc. That said, I would not like to do changes like these as a part of this PR, and would even consider doing it after the refactoring is merged, to avoid modification of business logic during the refactoring.

malinajirka replied:
Cool, I agree, thanks for recording it!

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/payment/controller/CardReaderTrackCanceledFlow.kt:
```diff
@@ -0,0 +1,41 @@
+                else -> null
+            }
+            if (nameForTracking == null) {
+                WooLog.e(WooLog.T.CARD_READER, "Invalid state received")
```
Same as above

malinajirka reviewed (approved):
Thanks for working on this @samiuelson !

The changes look good to me. I haven't run thorough tests since I believe we'll run them on the final PR of this chain of PRs anyway. I haven't discovered any issues when smoke testing the changes with my WisePad 3 (including interac refunds).

samiuelson commented:
Thanks for review, @malinajirka!

> I haven't run thorough tests since I believe we'll run them on the final PR of this chain of PRs anyway.

That's correct. The build used in the _Call for testing_ (pdfdoF-5Rg-p2) is generated from the next PR that includes payment state to UI state mapping — [[POS] Custom payment UI — Part 4 | Switching from viewStateData to paymentState](https://github.com/woocommerce/woocommerce-android/pull/12898).

samiuelson committed:
Rename `CardReaderTrackCanceledFlow` -> `CardReaderTrackCanceledFlowAction`
